src/fastpath/fastpath.c

```diff
@@ -6,15 +6,275 @@
+
+    /* Get destination thread VTable */
+    newVTable = TCB_PTR_CTE_PTR(dest, tcbVTable)->cap;
+
+    /* Ensure that the destination has a valid VTable. */
+    if (unlikely(!isValidVTableRoot_fp(newVTable))) {
+        slowpath(SysSend);
+    }
+
+    /* Get vspace root. */
+    cap_pd = cap_vtable_cap_get_vspace_root_fp(newVTable);
+
+#ifdef CONFIG_ARCH_AARCH32
+    /* Get HW ASID */
+    stored_hw_asid = cap_pd[PD_ASID_SLOT];
+#endif
+
+#ifdef CONFIG_ARCH_X86_64
+    /* borrow the stored_hw_asid for PCID */
+    stored_hw_asid.words[0] = cap_pml4_cap_get_capPML4MappedASID_fp(newVTable);
+#endif
+
+#ifdef CONFIG_ARCH_IA32
+    /* stored_hw_asid is unused on ia32 fastpath, but gets passed into a function below. */
+    stored_hw_asid.words[0] = 0;
+#endif
+
+#ifdef CONFIG_ARCH_AARCH64
+    stored_hw_asid.words[0] = cap_vtable_root_get_mappedASID(newVTable);
+#endif
+
+#ifdef CONFIG_ARCH_RISCV
+    /* Get HW ASID */
+    stored_hw_asid.words[0] = cap_page_table_cap_get_capPTMappedASID(newVTable);
+#endif
+
+    // TODO: Needed?
+    /* Try and wake up threads - this makes everything fail */
+    // awaken();
+
+    /* Let gcc optimise this out for 1 domain */
+    dom = maxDom ? ksCurDomain : 0;
+    /* Ensure only the idle thread or lower prio threads are present in the scheduler */
+    if (unlikely(dest->tcbPriority < NODE_STATE(ksCurThread->tcbPriority) &&
+                 !isHighestPrio(dom, dest->tcbPriority))) {
+
+        slowpath(SysSend);
+    }
+
+#ifdef CONFIG_ARCH_AARCH32
+    if (unlikely(!pde_pde_invalid_get_stored_asid_valid(stored_hw_asid))) {
+        slowpath(SysSend);
+    }
+#endif
+    /* Ensure the destination thread is in the current domain and can be scheduled directly. */
+    if (unlikely(dest->tcbDomain != ksCurDomain && maxDom)) {
+        slowpath(SysSend);
+    }
+#ifdef ENABLE_SMP_SUPPORT
+    /* Ensure both threads have the same affinity */
+    if (unlikely(NODE_STATE(ksCurThread)->tcbAffinity != dest->tcbAffinity)) {
+        slowpath(SysSend);
+    }
+#endif /* ENABLE_SMP_SUPPORT */
+    // TODO: Needed?
+    /* This is basically the check in checkDomainTime(), which needs more scheduler logic */
+    if (unlikely(isCurDomainExpired())) {
+        slowpath(SysSend);
+    }
+
+#ifdef CONFIG_BENCHMARK_TRACK_KERNEL_ENTRIES
+    ksKernelEntry.is_fastpath = true;
+#endif
+
+    thread_state_ptr_set_tsType_np(&dest->tcbState, ThreadState_Running);
+    updateTimestamp();
+    ticks_t prev = getNextInterrupt();
+
+    if (checkBudget()) {
+        commitTime();
+    }
+
+    if (isSchedulable(NODE_STATE(ksCurThread))) {
+        SCHED_ENQUEUE_CURRENT_TCB;
+    }
+
+    switchToThread_fp(dest, cap_pd, stored_hw_asid);
+
+    ticks_t next = getNextInterrupt();
+    if (next < prev) {
+        setDeadline(next - getTimerPrecision());
+    }
+
+    // TODO: Something like this is done in the slowpath - unsure if needed in fastpath.
+    if (sc_sporadic(dest->tcbSchedContext)) {
```

Comment:
The behaviour has to be equal, so the test to leave something out is if you can prove to yourself that the condition must be always false on the fastpath.